### PR TITLE
Yoda Yoda is

### PR DESCRIPTION
**Pull request by dkorolev**
File split + minor refactoring + got rid of one runtime exception my making it a compile-time check.

### CODE REVIEW

**Comment by dkorolev** [comment]
@mzhurovich : And now it's good to look at. Thanks!

File: yoda/types.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#include <type_traits>
+
+namespace yoda {
+
+// Entry key type extractor.
+template <typename T_ENTRY>
+using ENTRY_KEY_TYPE = typename std::remove_cv<
+    typename std::remove_reference<decltype(std::declval<T_ENTRY>().key())>::type>::type;
+
+// Base structures for user to signal that their entries should have or do emit certain properties.
```
Comment (mzhurovich) [inline yoda/types.h]: Maybe somehow mention that user is supposed to derive his entry types from one or several of these structures?

File: yoda/types.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+struct AllowNonThrowingGet {
+  constexpr static bool allow_nonthrowing_get = true;
+};
+
```
Comment (mzhurovich) [inline yoda/types.h:42]: ++ AllowOverwriteOnAdd

File: yoda/types.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+enum NullEntryTypeHelper { NullEntry };
+struct Nullable {
```
Comment (mzhurovich) [inline yoda/types.h:58]: Comment?

File: yoda/yoda.h [hunk 4]
```diff
@@ -23,126 +23,52 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
...
-//      Best for HTTP endpoints, with the Request object passed into these callbacks.
-//      `AsyncAdd(entry, on_success, on_failure)` and `AsyncGet(key, on_success, on_failure)`, both return void.
+//   1) Asynchronous, using std::{promise/future}, that utilize the the async/await paradigm.
+//      For instance, `AsyncAdd(entry)` and `AsyncGet(key)`, which both return a future.
 //
-//   3) Synchronous. `Get(key)` and `Add(entry)`.
+//   2) Asynchronous, using std::function as callbacks. Best for HTTP endpoints and other thread-based usecases.
+//      For instance, `AsyncAdd(entry, on_success, on_failure)` and `AsyncGet(key, on_success, on_failure)`,
+//      which both return `void` immediately and invoke the callback as the result is ready, from a separate
+//      thread.
```
Comment (mzhurovich) [inline yoda/yoda.h:39]: I'm not sure that I understand the sentence right, but we've just removed std::async()'ed callbacks, haven't we?

File: yoda/yoda.h [hunk 5]
```diff
@@ -151,45 +77,40 @@ T_ENTRY CreateNullEntry() {
...
  public:
   typedef ENTRY T_ENTRY;
+  typedef POLICY T_POLICY;
+
   typedef ENTRY_KEY_TYPE<ENTRY> T_KEY;
+
+  typedef sherlock::StreamInstance<T_ENTRY> T_STREAM_TYPE;
+
+  template <typename F>
+  using T_STREAM_LISTENER_TYPE = typename sherlock::StreamInstanceImpl<T_ENTRY>::template ListenerScope<F>;
```
Comment (mzhurovich) [inline yoda/yoda.h:108]: :+1: I already thought about defining these types :)

File: yoda/yoda.h [hunk 6]
```diff
@@ -199,26 +120,25 @@ class API {
...
+  std::future<T_ENTRY> AsyncGet(const T_KEY& key) {
     std::promise<T_ENTRY> pr;
     std::future<T_ENTRY> future = pr.get_future();
     state_maintainer_.mq_.EmplaceMessage(new MQMessageGet(key, std::move(pr)));
     return future;
   }
 
-  void AsyncGet(const T_KEY&& key, T_ENTRY_CALLBACK on_success, T_KEY_CALLBACK on_failure = [](const T_KEY&) {
-  }) {
+  void AsyncGet(const T_KEY& key, T_ENTRY_CALLBACK on_success, T_KEY_CALLBACK on_failure) {
```
Comment (mzhurovich) [inline yoda/yoda.h:137]: No more default empty callback on failure?

File: yoda/yoda.h [hunk 7]
```diff
@@ -274,23 +194,28 @@ class API {
...
-          null_entry.set_key(key);
-          pr.set_value(null_entry);
-        } else {  // Throw.
-          pr.set_exception(std::make_exception_ptr(T_KEY_NOT_FOUND_EXCEPTION(key)));
+        } else {
+          // Promise semantics.
+          SetPromiseToNullEntryOrThrow<T_KEY,
+                                       T_ENTRY,
+                                       T_KEY_NOT_FOUND_EXCEPTION,
+                                       T_POLICY::allow_nonthrowing_get>::DoIt(key, pr);
```
Comment (mzhurovich) [inline yoda/yoda.h:218]: :+1: Cool solution!

File: yoda/types.h (hunk 1 above)
Comment (dkorolev) [inline yoda/types.h]: They don't necessarily have to as of now.

We can _make_ them, but I don't see why should it be necessary as of now. What do you think?

File: yoda/yoda.h [hunk 8]
```diff
@@ -274,23 +194,28 @@ class API {
...
+          // Callback semantics.
           on_success(cit->second);
         } else {
+          // Promise semantics.
           pr.set_value(cit->second);
         }
-      } else {             // Key not found.
-        if (on_failure) {  // Callback function defined.
+      } else {
+        // The entry has not been fonud.
```
Comment (mzhurovich) [inline yoda/yoda.h]: s/fonud/found/

File: yoda/types.h (hunk 2 above)
Comment (dkorolev) [inline yoda/types.h:42]: Good idea, added.

(It worries me now that this magic `allow_nonthrowing_get` name is in several disconnected places though. We'll need to make sure the tests extensively cover these usecases.

File: yoda/yoda.h [hunk 9]
```diff
@@ -23,126 +23,52 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
...
-//      to his entry class definition.
 //
-//   2) By default, `Add` and `AsyncGet` throw `KeyAlreadyExistsException` if
-//      entry with this key already exists. User can change this behavior by
-//      addding
+//      to the definition of the entry class.
+//
+//   2) `Add()` and `AsyncGet()` throw `KeyAlreadyExistsException` if an entry with this key already exists.
+//
+//      This behavior can be changed by adding thThe user can change this behavior by adding:
```
Comment (mzhurovich) [inline yoda/yoda.h]: Looks like two sentences merged :)

File: yoda/types.h (hunk 3 above)
Comment (dkorolev) [inline yoda/types.h:58]: Yes, just added.

File: yoda/yoda.h (hunk 4 above)
Comment (dkorolev) [inline yoda/yoda.h:39]: I was trying to refer to the `on_*` callbacks... how would you rephrase it?

File: yoda/yoda.h (hunk 6 above)
Comment (dkorolev) [inline yoda/yoda.h:137]: Yes, I gave it another thought and realized it's a bad idea.

If, say, there's an HTTP endpoint, and the user simply forgets the last parameter, those HTTP requests will get hanging forever. It's a hard to catch bug. Not good, better eliminate all the room for it altogether.

File: yoda/yoda.h (hunk 8 above)
Comment (dkorolev) [inline yoda/yoda.h]: :+1: good eye, fixed.

**Comment by dkorolev** [comment]
@mzhurovich : Thanks for the review!

Asked clarification questions on two of your comments, and fixed the rest. PTAL!

File: yoda/types.h (hunk 1 above)
Comment (mzhurovich) [inline yoda/types.h]: No, I don't speak about changing our approach. Just to specify, that there is a set of base structures that can be used as parents for user's entry class.

Comment (dkorolev) [inline yoda/types.h]: Ah, I see, good point. One minute ...

File: yoda/yoda.h (hunk 6 above)
Comment (mzhurovich) [inline yoda/yoda.h:137]: Agree.

**Comment by dkorolev** [comment]
@mzhurovich : PTAL :-)

File: yoda/yoda.h (hunk 4 above)
Comment (mzhurovich) [inline yoda/yoda.h:39]: Let's leave it as is. It's just my specific reaction :)

Comment (dkorolev) [inline yoda/yoda.h:39]: Alrighty :)

**Comment by mzhurovich** [comment]
Everything looks good!
Merging.

**Pull request merged**